# Operations Research: Models & Applications

## Simple LP Formulations

### Problem 1 - Product Mix

- Objective: Maximise revenue from items made. £700 for a desk and £900 for a table.
- Decision variables - x1 (number of desks) & x2 (number of tables)
- Constraint 1: 3600 wood units. Each desk needs 3 units and a table 5.
- Constraint 2: 1600 labour hours available. Desk needs 1hr and Tables 2.
- Constraint 3: 800 machine hours available. 50 mins for a desk and 20 for a table

System of equations:

```
  max  700x1 + 900x2  
  s.t.   3x1 +   5x2 <= 3600  
          x1 +   2x2 <= 1600  
        50x1 +  20x2 <= 48000  
          x1         >= 0  
                  x2 >= 0  
```

Transformed system of equations (to have only <= constraints and a minimisation objective): 

```
  min  -700x1 - 900x2  
  s.t. +  3x1 +   5x2 <= +3600  
       +   x1 +   2x2 <= +1600  
       + 50x1 +  20x2 <= +48000  
          x1          >=  0  
                   x2 >=  0  
```

Test one of the possible solvers.

In [4]:
! pulptest

ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss	 Test that logic put in place for deprecation handling of indexs works
.	 Testing 'indexs' param continues to work for LpVariable.dicts
	 Testing 'indexs' param continues to work for LpVariable.matrix
.	 Testing 'indices' argument works in LpVariable.dicts
	 Testing 'indices' param continues to work for LpVariable.matrix
.	 Testing invalid status
.	 Testing continuous LP solution - export dict
.	 Testing export dict for LP
.	 Testing e

In [1]:
from scipy.optimize import linprog

In [2]:
obj = [-700, -900]
lhs_ineq = [
    [3, 5],
    [1, 2],
    [50, 20]
]
rhs_ineq = [
    3600,
    1600,
    48000
]
bnd = [
    (0, float("inf")),  # Bounds of x1
    (0, float("inf"))]  # Bounds of x2


In [3]:
opt = linprog(
    c=obj, A_ub=lhs_ineq, b_ub=rhs_ineq,
    # A_eq=lhs_eq, b_eq=rhs_eq, 
    bounds=bnd,
    method="revised simplex")

opt

     con: array([], dtype=float64)
     fun: -789473.6842105263
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([  0.        , 336.84210526,   0.        ])
  status: 0
 success: True
       x: array([884.21052632, 189.47368421])

From the output, of `opt.x` for x1 = 884.2 and x2 = 189.4, we have an optimum of £789,473. These figures could be rounded down and used as a starting point.

### Problem 2 - Production and Inventory

Story: Over next 4 days, a product needs to be made to the following quantities ready for orders: 100, 150, 200 & 170.
Production costs for each day are: £9, £12, £10, £12. Products can be made
and stored for another day. This is an inventory cost of £1/day

- Objective: Minimise costs while meeting target outputs for a product.
- ending inventory (t) =  beginning inventory(t) + production(t) - sales(t), t -> the nth day in the scenario.
- x -> production quantity as day start, y -> ending inventory
- Objective function: `9x1 + 12x2 + 10x3 + 12x4 + y1 + y2 + y3 + y4`


System of equations:

```
                       *sales* 
             +    x1  -  100 = y1 
          y1 +    x2  -  150 = y2  
          y2 +    x3  -  200 = y3  
          y3 +    x4  -  170 = y4
```
`y1, y2, y3 and y4 must all be >= 0. Because we must at least fulfill orders.`

### Problem 3 - Personnel Scheduling

Story: Over next 5 days, need to have employees in store. Everyone works 5 days and then rests for 2.

- Objective: Minimise number of hired employees
- Demand for each day:
- M -> 110, T -> 80, W -> 150, T -> 30, F -> 70, S -> 160, S -> 120, 


System of equations:

```
min: x1 + x2 + x3 + x4 + x5 + x6 + x7. 
   where x is no of employess that start on day i for 5 days.

Constraints:
x1 + x4 + x5 + x6 + x7 >= 110 # Monday's requirement
x2 + x5 + x6 + x7 + x1 >=  80 # Tuesday's requirement
x3 + x6 + x7 + x1 + x2 >= 150 # Wednesday's requirement
x4 + x7 + x1 + x2 + x3 >=  30 # Thursday's requirement
x5 + x1 + x2 + x3 + x4 >=  70 # Friday's requirement
x6 + x2 + x3 + x4 + x5 >= 160 # Saturday's requirement
x7 + x3 + x4 + x5 + x6 >= 120 # Sunday's requirement
```

In [7]:
obj_3 = [1, 1, 1, 1, 1, 1, 1]
lhs_ineq_3 = [
    [-1, -0, -0, -1, -1, -1, -1],
    [-1, -1, -0, -0, -1, -1, -1],
    [-1, -1, -1, -0, -0, -1, -1],
    [-1, -1, -1, -1, -0, -0, -1],
    [-1, -1, -1, -1, -1, -0, -0],
    [-0, -1, -1, -1, -1, -1, -0],
    [-0, -0, -1, -1, -1, -1, -1]
]
rhs_ineq_3 = [
    -110,
    -80,
    -150,
    -30,
    -70,
    -160,
    -120
]
bnd_3 = [
    (0, float("inf")),  # Bounds of x1
    (0, float("inf")),  # Bounds of x2
    (0, float("inf")),  # Bounds of x3
    (0, float("inf")),  # Bounds of x4
    (0, float("inf")),  # Bounds of x5
    (0, float("inf")),  # Bounds of x6
    (0, float("inf")),  # Bounds of x7
] 

opt_3 = linprog(
    c=obj_3, A_ub=lhs_ineq_3, b_ub=rhs_ineq_3,
    # A_eq=lhs_eq, b_eq=rhs_eq, 
    bounds=bnd_3,
    method="revised simplex")

opt_3

     con: array([], dtype=float64)
     fun: 163.33333333333331
 message: 'Optimization terminated successfully.'
     nit: 9
   slack: array([ 0.        , 70.        ,  0.        , 26.66666667,  0.        ,
        0.        ,  0.        ])
  status: 0
 success: True
       x: array([ 3.33333333, 40.        , 13.33333333,  0.        , 13.33333333,
       93.33333333,  0.        ])

In [10]:
print(f"Optimial solution from Monday to Saturday is {[f'{each}' for each in opt_3.x]}")

Optimial solution from Monday to Saturday is ['3.3333333333333335', '40.0', '13.333333333333334', '0.0', '13.333333333333332', '93.33333333333333', '0.0']
